oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationProvider.java

```diff
@@ -85,29 +93,30 @@ public Authentication authenticate(Authentication authentication) throws Authent
 				(OAuth2AuthorizationCodeAuthenticationToken) authentication;
 
 		OAuth2ClientAuthenticationToken clientPrincipal = null;
+		RegisteredClient registeredClient = null;
 		if (OAuth2ClientAuthenticationToken.class.isAssignableFrom(authorizationCodeAuthentication.getPrincipal().getClass())) {
 			clientPrincipal = (OAuth2ClientAuthenticationToken) authorizationCodeAuthentication.getPrincipal();
-		}
-		if (clientPrincipal == null || !clientPrincipal.isAuthenticated()) {
+			registeredClient = clientPrincipal.getRegisteredClient();
+		} else if (authorizationCodeAuthentication.getPrincipal() instanceof String) {
```

Comment:
So far, the principal for the `OAuth2AuthorizationCodeAuthenticationToken` is either `OAuth2ClientAuthenticationToken` or null (and in that case, there's a client ID). So for now we can do:

```java
/**
 * Returns the client id
 *
 * @return the client id
 */
public String getClientId() {
	if (this.clientPrincipal != null && OAuth2ClientAuthenticationToken.class.isAssignableFrom(this.clientPrincipal.getClass())) {
		return (String) this.clientPrincipal.getPrincipal();
	} else {
		return this.clientId;
	}
}
```

But I'm a bit worried about what happens if the principal is something else than an `OAuth2ClientAuthenticationToken` ; for now we just return null.

Thoughts ?

Reply:
I think it makes sense to provide a getter for `clientId` given we already provide one for `code` and `redirectUri`. I think we should set the `clientId` in the constructor that accepts `Authentication clientPrincipal` to simplify `getClientId()` and the signature should be `public @Nullable String getClientId()`.

FYI, `OAuth2ClientAuthenticationToken` will change as soon as we add support for other client authentication methods, which will ultimately change `OAuth2AuthorizationCodeAuthenticationToken` and likely other parts in the code. I also feel that `OAuth2AuthorizationCodeAuthenticationToken` should probably accept `OAuth2ClientAuthenticationToken` instead of `Authentication` but let's leave it as-is for now.

The changes suggested will be fine for this iteration but like I said things will change soon enough.